docs : add rfc for fast scan (#5414)
pingcap/tiflash: docs/design/2022-07-19-tiflash-support-fast-mode.md

# TiFlash Support Fast Mode

- Author(s): [Yunyan Hong](http://github.com/hongyunyan)

## Introduction

> This RFC makes TiFlash provide a brand new mode called "fast scan", to provide better query performance with a small loss of data consistency and timeliness.

Reviewer: ```suggestion
This RFC adds a brand new feature to TiFlash called "fast scan mode", providing better query performance with a small loss of data consistency and timeliness.
```

## Motivation or Background

> Currently, the query results provided by TiFlash provide the guarantee of snapshot isolation and strong data consistency. Thus, TiFlash provides accurate query results (MVCC filtering will be performed). However, in some AP scenarios, users do not have very high requirements for query accuracy, and can tolerate a certain degree of accuracy loss. Therefore, we propose corresponding modes to support the needs of users in different scenarios. That is, for scenarios that require high-precision query results, we can use our TiFlash Normal mode (default mode) to ensure the accuracy of query results. While for scenarios where query results can tolerate a certain error, and higher query performance is desired, we can use our new TiFlash mode (Fast Scan) to obtain higher query performance.

Reviewer: ```suggestion
Currently, TiFlash guarantees snapshot isolation and strong data consistency, i.e. MVCC filtering will be performed. However, in some AP scenarios, users do not have very high requirements for query accuracy, and can tolerate a certain degree of accuracy loss. Therefore, we propose corresponding scan modes to support different user scenarios. That is, for scenarios that require high-precision query results (by default), user can use the normal scan mode. For scenarios where user can tolerate a certain error in the query results, user can use the new fast scan mode to obtain higher query performance.
```

The capabilities of Fast Scan will be divided into multiple phases for continuous iteration and optimization.
The query results of table scan in the first phase of Fast Scan are as follows:

> 1. Does not support filtering of multiple versions of data
> 2. Filter the data of the row where the filter delete operation is located (the del_mark in the row equals 1)

Reviewer: ```suggestion
1. Multiple version of the row will be all preserved and not filtered out

2. Deleted versions will be filtered out
```

## Product Behavior

> We added a session/global variable ```tiflash_fastscan``` in TiDB  to control whether TiFlash enables fast scan. In default ```tiflash_fastscan = false```.

Reviewer: ```suggestion
We will add a session / global variable `tiflash_fastscan` in TiDB to control whether to enable fast scan during the query. By default `tiflash_fastscan = false`.
```

> If ```tiflash_fastscan``` is enabled at the session level, all tables in TiFlash in this session will query in fast scan mode. If ```tiflash_fastscan``` is enabled at the global level, in all sessions connected after the global variable is set, all tables in TiFlash will query in fast scan mode.

Reviewer: ```suggestion
If `tiflash_fastscan` is enabled in the session, all tables in TiFlash in this session will be queried in fast scan mode. If `tiflash_fastscan` is enabled globally, in all new sessions after the global variable is set, all tables in TiFlash will be queried in fast scan mode.
```

We can query the variable with the following statement.

```sql
SHOW SESSION VARIABLE LIKE 'tiflash_fastscan'
SHOW GLOBAL VARIABLE LIKE 'tiflash_fastscan'
```

```sql
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| tiflash_fastscan | OFF   |
+------------------+-------+
```

## Detailed Design

> ### Read Data Flow

Reviewer: ```suggestion
### Simplified Read Data Flow
```

> The TableScan operator in Normal Mode includes the following steps:

Reviewer: ```suggestion
Currently, the TableScan process in the Normal Scan Mode includes the following steps:
```

1. Read data from Delta + Stable: Establish 3 data streams to read data from the MemTableSet and PersistedFileSet of the Delta layer, as well as the Stable layer.
2. Sort Merge: Merge the 3 data streams established in step 1, and return the data in the increasing order of (handle, version).
3. Handle Filter: Filter the data according to the requested handle ranges.
4. MVCC + Column Filter: filter the data by MVCC according to version and del_mark, and filter out the unnecessary internal columns to return the data.

> Considering that MVCC filtering is a time-consuming step, but the impact on statistical results is limited, in Fast Mode, we propose omitting the MVCC filtering step to speed up read operations. At the same time, because we omit the MVCC step, we also do not need to merge the data of the delta layer and the stable layer and return them in the order of (handle, version).

Reviewer: ```suggestion
MVCC filtering is a time-consuming step. In the Fast Scan Mode, **the MVCC filtering step will be omitted** to speed up read operations. As a result, we also **omit the step 2** above, i.e. no need to merge the data of the delta layer and the stable layer which returns in the order of (handle, version).
```

> Therefore, we can simplify the TableScan process in fast mode to (omit step 2 and MVCC filtering in step 4 of the Normal Mode)

Reviewer: ```suggestion
In conclusion, in the Fast Scan Mode, the TableScan process is as follows:
```

1. Read data from Delta + Stable: Establish three data streams to read data from the MemTableSet and PersistedFileSet of the Delta layer, as well as the Stable layer.
2. Handle Filter: Filter the data according to the data range.
3. Delete Filter: Filter the data according to del_mark (filter out the row with del_mark = 1), and filter out the unnecessary columns to return the data.

> 4. Merge : Merge and return the above data streams directly (no need to merge in order, just concat)

Reviewer: ```suggestion
4. Concat: Merge and return the above data streams directly (no need to merge in order, just concatenation).
```

Samples:
ALTER TABLE test.t SET TIFLASH MODE FAST;
ALTER TABLE test.t SET TIFLASH MODE NORMAL;
```

Currently TiFlash does not support temporary tables, system tables, memory tables, and tables with non-utf-8 characters in their column names, thus executing this statement on these tables will return an error.

Besides, we will add a field 'table_mode' to the system table information_schema.tiflash_replica to support the user viewing the TiFlash mode of the current table.

```sql
select * from information_schema.tiflash_replica
```

## Detailed Design

### DDL Implementation

The schema of the TiFlash table will be saved as part of the schema information of each table. In the table_info of the table, we will add a TiFlashMode field to identify the corresponding mode. This field will follow the sync schema approach among TiDB/TiKV/TiFlash as other schema infos.

```c++
enum class TiFlashMode
{
    Normal,
    Fast,
};
```

In order to facilitate checking mode, we will add the TABLE_MODE field to the system table information_schema.tiflash_replica to store the schema of the table.

### Read Data Flow

The TableScan operator in Normal Mode includes the following steps:

> 1. Read data from Delta + Stable: Establish three data streams to read data from the MemTableSet and PersistedFileSet of the Delta layer, as well as the Stable layer.
> 2. Sort Merge : Merge the 3 data streams established in step 1, and return the data in the order of (handle, version).
> 3. Handle Filter : Filter the data according to the data range.
> 4. MVCC + Column Filter: filter the data by MVCC according to version and del_mark, and filter out the unnecessary columns to return the data.

Reviewer 2: ```suggestion
1. Read data from Delta + Stable: Establish 3 data streams to read data from the MemTableSet and PersistedFileSet of the Delta layer, as well as the Stable layer.
2. Sort Merge: Merge the 3 data streams established in step 1, and return the data in the increasing order of (handle, version).
3. Handle Filter: Filter the data according to the handle ranges requested.
4. MVCC + Column Filter: filter the data by MVCC according to version and del_mark, and filter out the unnecessary internal columns to return the data.
```

> Considering that MVCC filtering is a time-consuming step, but the impact on statistical results is limited, in Fast Mode, we propose to omit the step of MVCC filtering to speed up read operations. At the same time, because we omit the MVCC step, we also do not need to merge the data of the delta layer and the stable layer and return them in the order of (handle, version).

Reviewer 2: ```suggestion
Considering that MVCC filtering is a time-consuming step, but the impact on statistical results is limited, in Fast Mode, we propose omitting the MVCC filtering step to speed up read operations. At the same time, because we omit the MVCC step, we also do not need to merge the data of the delta layer and the stable layer and return them in the order of (handle, version).
Note that the rows returned under Fast Mode are not in order, TiDB should not generate plans with "keep order" TableScan operator when accessing those TiFlash tables.
```
